Move all MACs to the providers

levitte opened the pull request:
Because EVP_MAC is a new API, there is no legacy code to keep around, and the API is entirely adapted to support provider based implementations.  This affects code in all kinds of places, all the way up to how `openssl list -mac-algorithms` does its work.

It's recommended to read the commit messages for explanation of some of the changes.

-----

This currently depends on #9303

mattcaswell commented on crypto/evp/p_lib.c:
```diff
@@ -328,12 +334,29 @@ EVP_PKEY *EVP_PKEY_new_CMAC_key(ENGINE *e, const unsigned char *priv,
         goto err;
     }
+#endif
+    /* How to support engines ? */
```
I think if the engine param is set then this has to be treated entirely separately. That is essentially an instruction to lookup the provided cipher in an engine provider (i.e. the EVP_MAC_fetch call does something like "engine=yes enginename=blah")

mattcaswell replied:
Hmmm....although on second thoughts its the cipher that needs to be looked up so it actually needs to be an EVP_CIPHER_fetch call somewhere - but perhaps that should be in the default provider. So maybe this approach of sending the ending name as a param is the right approach after all.

mattcaswell commented on crypto/evp/p_lib.c:
```diff
@@ -328,12 +334,29 @@ EVP_PKEY *EVP_PKEY_new_CMAC_key(ENGINE *e, const unsigned char *priv,
+                                         strlen(engine_name) + 1, NULL);
+    params[paramsn++] =
+        OSSL_PARAM_construct_utf8_string("cipher", (char *)cipher_name,
+                                         strlen(cipher_name) + 1, NULL);
```
This needs some thought. The cipher might be an explicitly fetched cipher in which case we should be using that specific cipher.

levitte committed:
Prepare EVP_MAC infrastructure for moving all MACs to providers

Quite a few adaptations are needed, most prominently the added code
to allow provider based MACs.

As part of this, all the old information functions are gone, except
for EVP_MAC_name().  Some of them will reappear later, for example
EVP_MAC_do_all() in some form.

MACs by EVP_PKEY was particularly difficult to deal with, as they
need to allocate and deallocate EVP_MAC_CTXs "under the hood", and
thereby implicitly fetch the corresponding EVP_MAC.  This means that
EVP_MACs can't be constant in a EVP_MAC_CTX, as their reference count
may need to be incremented and decremented as part of the allocation
or deallocation of the EVP_MAC_CTX.  It may be that other provider
based EVP operation types may need to be handled in a similar manner.

levitte committed:
Move BLAKE2 MACs to the providers

This also moves the remaining parts of BLAKE2 digests to the default
provider, and removes the legacy EVP implementation.

levitte committed:
Move CMAC to providers

levitte committed:
Move GMAC to providers

levitte committed:
Move HMAC to providers

levitte committed:
Move KMAC to providers

Instead of using evp_keccak_kmac128() and evp_keccak_kmac256(), we refer
to the hash implementation by name, and fetch it, which should get us the
implementation from providers/common/digests/sha3_prov.c.

levitte committed:
Move SipHash to providers

levitte committed:
Move Poly1305 to providers

levitte committed:
Remove init of MACs from EVP

Now that all our MACs have moved to the default provider, we let it
take over completely

levitte committed:
Adapt apps/mac.c to use provider based MACs

levitte committed:
Adapt diverse code to provider based MACs.

CRMF, SSKDF, TLS1_PRF and SIV are affected by this.

This also forces the need to check MAC names, which leads to storing
the names in the created methods, which affects all EVP APIs, not just
EVP_MAC.  We will want that kind of information anyway (for example
for 'openssl list')...  Consequently, EVP_MAC_name() is re-implemented.

levitte committed:
Add EVP_MAC_provider()

For information processing.

levitte committed:
Implement EVP_MAC_do_all_ex()

levitte commented:
I've put down quite a lot of work in this, and there are multiple parts this change affects.

The EVP_MAC API itself
====================

Because the EVP_MAC API is completely new and unreleased, I chose to throw away all the legacy stuff and make this completely provider based.

Most notably, this means that MACs are *always* fetched and that control functions are replaced with parameter passing functions.
It also means that MACs have no presence in the OBJ databases, except for the cases where a legacy EVP_PKEY implementation exists.

Information services
================

`openssl list -mac-algorithms` needed new code to be able to display information on existing MAC algorithms, and since the OBJ based functionality is now gone, a re-implementation of EVP_MAC_do_all and EVP_MAC_do_all_sorted is now needed.

TODO
=====

- Parameter names are currently scattered all over the code and need to be collected as macros in `include/openssl/core_names.h`.
- Some functions and libcrypto\<-\>provider interfaces are not yet implemented.

levitte commented:
A note on these functions:

``` C
const OSSL_PARAM *EVP_MAC_CTX_get_param_types(const EVP_MAC *mac);
const OSSL_PARAM *EVP_MAC_CTX_set_param_types(const EVP_MAC *mac);
```

According to design discussions, these should return OSSL_ITEM arrays.  However, that type is poor in information, especially regarding expected value sizes, and rather than creating yet another type, I thought that re-using `OSSL_PARAM` for parameter declaration and discovery purposes could be a viable idea.  Special semantics apply to the `data_size` field, which means "arbitrary length" in this use.

levitte commented:
I would appreciate some commentary on this

beldmit commented:
Current TLS implementation refers to MAC NIDs. How can we change it to avoid this reference for future algorithms?

levitte commented:
> Current TLS implementation refers to MAC NIDs. How can we change it to avoid this reference for future algorithms?

Refer to them by name and use `EVP_MAC_fetch()`

beldmit commented:
I mean places similar to ssl_cipher_get_evp() and ssl_load_ciphers(). So for TLS < 1.3 we either have to refactor this code or register NIDs for new primitives. It seems not a problem but should be taken into account.

mattcaswell commented on providers/common/macs/cmac.c:
```diff
@@ -0,0 +1,210 @@
+/*
+ * TODO(3.0) solve the ENGINE problem
+ *
+ * For the moment, ENGINEs are simply ignored.  A tentative idea is to
```
I don't think we can just ignore this problem. We shouldn't be introducing temporary code breaks. We should instead fix the underlying problem first.

levitte replied:
Like I said elsewhere, I've started working in ENGINE by name.  This comment will disappear.

levitte replied:
Comment removed, engines are now handled here.

mattcaswell commented on providers/common/macs/cmac.c:
```diff
@@ -0,0 +1,210 @@
+static const OSSL_PARAM known_params[] = {
+    OSSL_PARAM_utf8_string(OSSL_MAC_PARAM_ALGORITHM, NULL, 0),
+    OSSL_PARAM_utf8_string(OSSL_MAC_PARAM_CIPHER, NULL, 0),     /* Same as "algorithm" */
+    OSSL_PARAM_utf8_string(OSSL_MAC_PARAM_PROPERTIES, NULL, 0), /* instead of ENGINE */
```
Where is `OSSL_MAC_PARAM_PROPERTIES` defined...I don't see it in core_names...or am I going blind?

levitte replied:
Fixing

mattcaswell commented on doc/man3/EVP_MAC.pod:
```diff
@@ -87,179 +101,168 @@ through diverse controls.
+
+=item "cipher"
+
+=item "algorithm"
```
The "properties" parameter should be described here

levitte replied:
I have removed property params for now, need to have another look at surrounding code.  They will come back.

mattcaswell commented on providers/common/macs/cmac.c:
```diff
@@ -0,0 +1,210 @@
+                EVP_CIPHER_meth_free(macctx->alloc_cipher);
+
+            if ((macctx->tmpcipher = macctx->alloc_cipher =
+                 EVP_CIPHER_fetch(NULL, algoname, properties)) == NULL) {
```
`libctx` should not be NULL. This is a bit of a problem actually. In the FIPS provider it should be the provider context that was passed as the parameter to `cmac_new` (which is really an OPENSSL_CTX). But in the default provider we should use the OPENSSL_CTX from the libcrypto side. Which makes me realise that this FIPS_MODE only code should perhaps apply to any built-in provider (or perhaps just all providers):

https://github.com/openssl/openssl/blob/85e0ad9f1309afceaa8d3946c9273e089d3c25b8/crypto/provider_core.c#L402-L424

In other words in the default provider, the provider context should be the same object as the "real" OPENSSL_CTX.

levitte replied:
I would say that any provider is free to use whatever context and whatever properties they want, so let's not push this on *everyone*.

The idea with the provider context is that the provider decides what it contains, not libcrypto, so forcing that to be a `OPENSSL_CTX` at all times is not a good idea either.

Now, what *we* choose to do with *our* providers is a different matter, and of course we can make whatever we choose to use available for others as well, but that's the point, isn't it, that it should remain a choice?
Also, I'm apprehensive of "too magic", so I'm wondering if things can get redesigned so the provider can *ask* for the library context that it's running under...  as an upcall with the `OSSL_PROVIDER` as argument, maybe?

levitte replied:
Also, note that this PR is a draft, so some things are obviously lacking, and I wanted to see that the code would build at all, and even test.

mattcaswell replied:
> I would say that any provider is free to use whatever context and whatever properties they want, so let's not push this on everyone.

> The idea with the provider context is that the provider decides what it contains, not libcrypto, so forcing that to be a OPENSSL_CTX at all times is not a good idea either.

I'm definitely not suggesting that. My only suggestion was what the provctx value is by default. Any provider would be free to set it to whatever they want. This is a `void *` so there is nothing forcing an OPENSSL_CTX to be used. Currently it defaults to NULL, except in the special case of the built-in internal recursive FIPS provider where it defaults to point to the "current" OPENSSL_CTX. I'm just saying we can make that the default everywhere.

> Also, I'm apprehensive of "too magic", so I'm wondering if things can get redesigned so the provider can ask for the library context that it's running under... as an upcall with the OSSL_PROVIDER as argument, maybe?

Yes, that could work too.

levitte replied:
> Yes, that could work too.

Cool, I'll look into that next week.

levitte replied:
I have changed them all to use the library context that the provider works in.

mattcaswell commented on crypto/evp/p_lib.c:
```diff
@@ -318,8 +319,13 @@ EVP_PKEY *EVP_PKEY_new_CMAC_key(ENGINE *e, const unsigned char *priv,
+    const char *cipher_name = EVP_CIPHER_name(cipher);
     EVP_PKEY *ret = EVP_PKEY_new();
-    EVP_MAC_CTX *cmctx = EVP_MAC_CTX_new_id(EVP_MAC_CMAC);
+    EVP_MAC *cmac = EVP_MAC_fetch(NULL, "CMAC", NULL);
```
Should we use the same libctx that the cipher was fetched from (if it was)?

levitte replied:
That would be sensible

levitte replied:
Done, will show up in a commit soon

levitte replied:
[In the autumn?](https://gist.github.com/levitte/5cbabd94b19f1a022fdd1c2f4442e2b7)

...

or I can have a look now...

levitte replied:
Done

mattcaswell commented on providers/common/macs/gmac.c:
```diff
@@ -0,0 +1,214 @@
+                EVP_CIPHER_meth_free(macctx->alloc_cipher);
+
+            if ((macctx->cipher = macctx->alloc_cipher =
+                 EVP_CIPHER_fetch(NULL, algoname, properties)) == NULL) {
```
Shouldn't use NULL here

mattcaswell commented on providers/common/macs/gmac.c:
```diff
@@ -0,0 +1,214 @@
+
+struct gmac_data_st {
+    const EVP_CIPHER *cipher;    /* Cache GCM cipher */
+    const ENGINE *engine;        /* Engine implementating the algorithm */
```
Doesn't seem to be used?

levitte replied:
I will remove

levitte replied:
Oh wait, I did add use of it!

mattcaswell commented on providers/common/macs/hmac.c:
```diff
@@ -0,0 +1,205 @@
+struct hmac_data_st {
+    /* tmpmd and tmpengine are set to NULL after a HMAC_Init call */
+    const EVP_MD *tmpmd;         /* HMAC digest */
+    const ENGINE *tmpengine;     /* HMAC digest engine */
```
Not set anywhere

levitte replied:
Is now!

mattcaswell commented on providers/common/macs/hmac.c:
```diff
@@ -0,0 +1,205 @@
+    HMAC_CTX *ctx;               /* HMAC context */
+};
+
+static void *hmac_new(void *provctx)
```
provctx needs to be saved away for future fetch operations

mattcaswell commented on providers/common/macs/hmac.c:
```diff
@@ -0,0 +1,205 @@
+                EVP_MD_meth_free(macctx->alloc_md);
+
+            if ((macctx->tmpmd = macctx->alloc_md =
+                 EVP_MD_fetch(NULL, algoname, properties)) == NULL) {
```
Shouldn't be NULL

levitte commented on crypto/evp/evp_lib.c:
```diff
@@ -700,3 +700,117 @@ int EVP_hex2ctrl(int (*cb)(void *ctx, int cmd, void *buf, size_t buflen),
+
+                    c ^= 255;
+                    c += carry;
+                    carry = c == 0;
```
Ah, good point

levitte commented on crypto/evp/evp_lib.c:
```diff
@@ -700,3 +700,117 @@ int EVP_hex2ctrl(int (*cb)(void *ctx, int cmd, void *buf, size_t buflen),
+                    carry = c == 0;
+                    *cp = c;
+                }
+            }
```
Oh! Good idea!

levitte commented on include/openssl/evp.h:
```diff
@@ -1001,53 +1002,33 @@ void EVP_MD_do_all_sorted(void (*fn)
+                    void *arg);
+void EVP_MAC_do_all_sorted(OPENSSL_CTX *libctx,
+                           void (*fn)(const EVP_MAC *mac, void *arg),
+                           void *arg);
```
Have a look at the change in `apps/openssl.c`

levitte replied:
(we seem to forget those little tid-bits, such as presentation...  I'm trying to get us to remember those bits)

levitte commented on doc/man3/EVP_MAC.pod:
```diff
@@ -60,25 +69,30 @@ B<EVP_MAC_CTX> is a context type that holds internal MAC information
+=head2 Algorithm implementation fetching
+
+EVP_MAC_fetch() fetches an implementation of a MAC I<algorithm>, given
+a library context I<libctx> and a set of I<properties>.
```
Yeah, B is our own invention. I've had a look at the standards for writing man-pages, and they go for I. 

See http://man7.org/linux/man-pages/man7/man-pages.7.html

mattcaswell replied:
Which suggests we need a PR to change all our existing man pages! Ouch.

levitte replied:
Yes.  Someone did submit such a PR a long while ago, I'm afraid it got lost in all the noise, and so much has changed since that it really needs to be redone.  Most of it can be automated, and I'm willing to do the work.

richsalz replied:
What about a PR to doc-nits and fixes that catch their list of "do not use" phrases?  I can do that if desired.

levitte replied:
"do not use" phrases?  Not sure what you're after, but it's still allowed to use Both `B<` and `I<` for emphasis, so it's a bit difficult to know the author's intent.
(emphasis should probably not be needed in manuals, though)

levitte commented on doc/man3/EVP_MAC.pod:
```diff
@@ -60,25 +69,30 @@ B<EVP_MAC_CTX> is a context type that holds internal MAC information
+
+EVP_MAC_fetch() fetches an implementation of a MAC I<algorithm>, given
+a library context I<libctx> and a set of I<properties>.
+
```
Sure

levitte commented on doc/man3/EVP_MAC.pod:
```diff
@@ -87,179 +101,168 @@ through diverse controls.
 Some MACs do not support this option.
 
-=item B<EVP_MAC_CTRL_SET_ENGINE>
+=item "engine"
```
Not sure what you mean.

levitte replied:
Ah, now I get it.  They are described as a group, so it looks like this:

>        "engine"
>        "md"
>        "digest"
>        "cipher"
>        "algorithm"
>            For MAC implementations that use an underlying computation
>            algorithm, these parameters set what the algorithm should be, and
>            the engine that implements the algorithm if needed.
> 
>            These parameter should all be of type OSSL_PARAM_UTF8_STRING, and
>            the value is always the name of the intended engine or algorithm.
> 
>            Note that not all algorithms may support all digests.  HMAC does
>            not support variable output length digests such as SHAKE128 or
>            SHAKE256.
> 
>            Also note that "algorithm" can be use generically instead of "md",
>            "digest" or "cipher", and that "md" and "digest" are also
>            interchangable.

mattcaswell replied:
Oh right...that makes sense.

levitte commented on apps/mac.c:
```diff
@@ -109,7 +92,7 @@ int mac_main(int argc, char **argv)
     }
 
-    mac = EVP_get_macbyname(argv[0]);
+    mac = EVP_MAC_fetch(NULL, argv[0], NULL);
```
Will fix

levitte commented on util/libcrypto.num:
```diff
@@ -4576,21 +4576,21 @@ EVP_MD_CTX_set_pkey_ctx                 4531	3_0_0	EXIST::FUNCTION:
+EVP_MAC_str2ctrl                        4546	3_0_0	NOEXIST::FUNCTION:
+EVP_MAC_hex2ctrl                        4547	3_0_0	NOEXIST::FUNCTION:
+EVP_MAC_nid                             4548	3_0_0	NOEXIST::FUNCTION:
+EVP_get_macbyname                       4549	3_0_0	NOEXIST::FUNCTION:
```
Yeah well, that will leave a hole, which has the exact same effect.  Question is if we should rebuild libcrypto.num from scratch before the 3.0 release, just like we did with the 1.1.0 release...

mattcaswell replied:
> Yeah well, that will leave a hole, which has the exact same effect.

Unless you just delete everything below this point and re-run `make update`.

> if we should rebuild libcrypto.num from scratch before the 3.0 release

Yeah, we could do that.

levitte replied:
> Unless you just delete everything below this point and re-run `make update`.

That involves much more than this PR, I would prefer that to happen in a separate cleanup PR.

levitte commented on test/evp_test.c:
```diff
@@ -1135,97 +1139,67 @@ static int mac_test_run_mac(EVP_TEST *t)
+    /*
+     * Unknown controls.  They must match parameters that the MAC recognises
+     */
+    if (params_n + sk_OPENSSL_STRING_num(expected->controls) > 20) {
```
Yes.  Will fix.

levitte commented on providers/default/defltprov.c:
```diff
@@ -115,6 +115,20 @@ static const OSSL_ALGORITHM deflt_ciphers[] = {
+    { "KMAC128", "default=yes", kmac128_functions },
+    { "KMAC256", "default=yes", kmac256_functions },
+    { "SipHash", "default=yes", siphash_functions },
+    { "Poly1305", "default=yes", poly1305_functions },
```
Will fix.

levitte commented on crypto/evp/evp_lib.c:
```diff
@@ -700,3 +700,117 @@ int EVP_hex2ctrl(int (*cb)(void *ctx, int cmd, void *buf, size_t buflen),
+int EVP_ctrl_str_to_param(OSSL_PARAM *to,
+                          const char *type,
+                          const char *value, size_t value_n,
+                          const OSSL_PARAM *paramdef)
```
Yes, and that isn't the only one

mattcaswell replied:
We really should extend make doc-nits to complain about newly added functions that don't have documentation.

levitte replied:
Feel free to do that.  There *is* a mechanism to check how many undocumented functions there are:

```
: ; ./util/find-doc-nits -u
# Duplicate i2d_PublicKey in doc/man3/d2i_X509.pod and doc/man3/d2i_PrivateKey.pod
# Found 4826 in util/libcrypto.num
# Found 1497 missing from util/libcrypto.num

# Found 506 in util/libssl.num
# Found 44 missing from util/libssl.num

# Checking macros (approximate)
# Found 215 macros missing (not all should be documented)
```

What we need is a way for it to know what's "new"

mattcaswell replied:
> What we need is a way for it to know what's "new"

That should just be a case of capturing a list of everything that is currently undocumented and freezing it. We then just make sure that everything reported by `find-doc-nits -u` is on that list. If it isn't then its a newly added undocumented function.

richsalz replied:
See https://github.com/openssl/openssl/issues/8798 for discussion about this.

levitte commented on crypto/evp/evp_lib.c:
```diff
@@ -700,3 +700,117 @@ int EVP_hex2ctrl(int (*cb)(void *ctx, int cmd, void *buf, size_t buflen),
+             */
+            BIGNUM *a = NULL;
+
+            BN_dec2bn(&a, value);
```
We have never use hex for anything other octet strings (or well, what they are in pre-3.0) as far as I know.  But sure, that can be arranged.

levitte commented on crypto/evp/evp_lib.c:
```diff
@@ -700,3 +700,117 @@ int EVP_hex2ctrl(int (*cb)(void *ctx, int cmd, void *buf, size_t buflen),
+            /*
+             * BN_bn2nativepad puts the absolute value of the number in
+             * the buffer, i.e. if it's negative, we need to deal with it.
+             * We do *not* care about overflows, i.e. the size stays the
```
We do pass signed numbers to the controls, so there's the *potential*, even though we have made no explicit support.  The BIGNUM tests include negative numbers, so we are prepared for the possibility.  So I guess this comes down to being "feature complete" or just doing the minimum for what we think will come today.

levitte commented on crypto/core_algorithm.c:
```diff
@@ -0,0 +1,62 @@
+#include <openssl/core.h>
+#include "internal/core.h"
+#include "internal/property.h"
+#include "internal/provider.h"
```
Sure.  I found that `EVP_MAC_do_all` would do very much what `EVP_MAC_fetch`, so instead of creating something that I knew would be copied to any other API where we care about presentation, I'd make something that's general purpose enough to code the major part of the job and make the like of `EVP_MAC_do_all` as well as `EVP_MAC_fetch` easy to write.

This needs documenting, btw...

mattcaswell replied:
Ah. Thanks for the explanation.

levitte commented on crypto/evp/c_alld.c:
```diff
@@ -46,10 +46,6 @@ void openssl_add_all_digests_int(void)
-#endif
-#ifndef OPENSSL_NO_BLAKE2
-    EVP_add_digest(EVP_blake2b512());
-    EVP_add_digest(EVP_blake2s256());
```
I wonder too...  I think that's an early experiment that eventually lead to #9075, and that I forgot to clean away from here.  Will clean.

levitte commented on crypto/evp/p_lib.c:
```diff
@@ -328,12 +334,30 @@ EVP_PKEY *EVP_PKEY_new_CMAC_key(ENGINE *e, const unsigned char *priv,
         goto err;
     }
 
+#if 0
```
Will clean

levitte replied:
Cleaned away now.

levitte commented on crypto/evp/pkey_mac.c:
```diff
@@ -187,33 +191,41 @@ static int pkey_mac_signctx_init(EVP_PKEY_CTX *ctx, EVP_MD_CTX *mctx)
+        size_t params_n = 0;
+        int flags = EVP_MD_CTX_test_flags(mctx, ~EVP_MD_CTX_FLAG_NO_INIT);
+
+        /* TODO(3.0) "flagss" isn't quite right, i.e. a quick hack for now */
```
Okie

levitte commented on crypto/evp/pkey_mac.c:
```diff
@@ -228,14 +240,23 @@ static int pkey_mac_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
+                size_t params_n = 0;
+                char *ciphname = (char *)OBJ_nid2sn(EVP_CIPHER_nid(p2));
+
+#if 0                            /* TODO(3.0) figure out ENGINE as param */
```
Yeah...  I've recently decided to go for engine *names*, i.e. let the receiving end get the actual ENGINE

levitte replied:
Fixed

levitte commented on crypto/kdf/sskdf.c:
```diff
@@ -327,7 +346,11 @@ static int sskdf_ctrl(EVP_KDF_IMPL *impl, int cmd, va_list args)
-        mac = va_arg(args, const EVP_MAC *);
+        /*
+         * TODO(3.0) instead of passing a EVP_MAC, an name should be passed,
+         * and the EVP_MAC instance should be created here.
```
Good question.

The reason for this comment is that I expect KDFs to be moved entirely to providers as well, and passing a EVP_MAC across the libcrypto\<-\>provider boundary isn't acceptable, so...  I'm thinking that the same mechanism as for MAC algorithms, i.e. pass a MAC name, and provide for passing properties as well.

levitte commented on crypto/crmf/crmf_pbm.c:
```diff
@@ -193,17 +199,22 @@ int OSSL_CRMF_pbm_new(const OSSL_CRMF_PBMPARAMETER *pbmp,
             || !EVP_MAC_init(mctx)
             || !EVP_MAC_update(mctx, msg, msglen)
-            || !EVP_MAC_final(mctx, mac_res, maclen))
+        || !EVP_MAC_final(mctx, mac_res, outlen, EVP_MAX_MD_SIZE))
```
Fixed

levitte commented:
TODO
====

- Figure out OPENSSL_CTX passing (will be another PR, as this should be solved on a wider scale)
- Figure out passing of properties (just pass as a utf8 string parameter, I presume)

levitte commented:
OPENSSL_CTX passing was pretty much solved in #9160.  I need to work a little bit more on this PR to integrate that, but it's getting close.

It currently builds, at least on my machine, and tests pretty fine, except for the BLAKE2 MACs...

levitte commented:
**This is no longer a draft**

levitte commented:
Argh, of course, the OSSL_PARAM return size change makes it all go belly up...

It might take a week before I get to this...

paulidale commented:
Apologies :(

levitte commented:
I approved it, so it's as much on me.

levitte commented:
Hah, did find some time :-)

levitte commented:
Oh wow, Travis still stuck?

levitte commented:
Another rebase.

levitte commented:
> I mean places similar to ssl_cipher_get_evp() and ssl_load_ciphers(). So for TLS \< 1.3 we either have to refactor this code or register NIDs for new primitives. It seems not a problem but should be taken into account.

I agree that those adaptations will have to happen.  Is that something you'd like to see as part of this PR or can it be left for another?

levitte commented:
`EVP_ctrl_str_to_param` is really not just a EVP thing, it should be an `OSSL_PARAM` function.  I'm preparing a separate PR for it.

levitte commented:
I amended the description to point out the dependence on #9303

levitte commented:
I'm thinking that I should break out the apps specific parts to another PR.  That might make this one easier to digest.

levitte committed:
Adapt the MAC tests, and tests for other things that use EVP_MAC

levitte commented:
This depends on #9303, which is code that was originally here but got generalised

levitte committed:
Re-implement 'openssl list -mac-algorithms'

levitte committed:
Adjust some provider reason codes

BLAKE2 MACs came with a set of new reason codes.  Those talking about
lengths are consistently called PROV_R_INVALID_FOO_LENGTH, for any
name FOO.  The cipher messages were briefer.  In the interest of
having more humanly readable messages, we adjust the reasons used by
the ciphers (that's just IV length and key length).

levitte committed:
OSSL_PARAM_construct_from_text(): handle non-hex octet string input

levitte commented:
All dependencies are cleared, time to have a final look at this.

paulidale commented on test/evp_test.c:
```diff
@@ -1251,6 +1229,9 @@ static int mac_test_run_mac(EVP_TEST *t)
     }
     t->err = NULL;
  err:
+    for (; params_n-- > params_n_allocstart;) {
```
Isn't this just a `while` loop?

levitte replied:
True.  Will fix

paulidale commented on providers/fips/fipsprov.c:
```diff
@@ -274,6 +274,12 @@ static const OSSL_ALGORITHM fips_ciphers[] = {
 
+static const OSSL_ALGORITHM fips_macs[] = {
+    { "GMAC", "default=yes", gmac_functions },
+    { "HMAC", "default=yes", hmac_functions },
```
NIST has standards for CMAC and KMAC.  The two KMACs are up above in the digests, I thought it was plumbed in via MAC.  @slontis you did this didn't you?

CMAC is used as part of other things (memory fails me as to what) and is more common than GMAC.  KMAC isn't common yet.

Even if extra algorithms are included in the FIPS provider, they don't need to be part of the validation.  Non-approved algorithms are allowed and get documented as such.  I kind of suspect that the lab isn't going to object to any extras, they'll have to do validations for them as well in due course, so no effort gets saved.  I've been meaning to ask about this.

levitte replied:
I'm surprised you didn't comment on the utterly faulty property! :wink:

levitte replied:
Regarding the use as MD or as MAC, I'm honestly utterly confused.  What @slontis has implemented is a keyed hash, and sure, that's essentially what a MAC is...  but how do we want to present that to the users, as a `EVP_MD` or as a `EVP_MAC`?  We're currently doing both.

levitte replied:
Regarding the choice of MACs to be included in the FIPS provider module, I relied on the choices made in https://www.openssl.org/docs/OpenSSL300Design.html, see Appendix 3.

levitte replied:
Meanwhile, I fixed the property strings

paulidale replied:
Yet, KMAC is included.  I still think that there is some merit in including them in the FIPS module even if they aren't immediately validated.  It would make validating them easier and we might get lucky with the lab :)

@slontis, thoughts about the MD vs MAC for KMAC?

slontis replied:
Well its name kind of gives it away :) It should be a MAC.

levitte replied:
So why do we have it implemented as a hash?

levitte replied:
Specifically, these two lines confused the hell out of me:

https://github.com/openssl/openssl/blob/697b0c5185f2b379cf23330fddc5f8b2d691db17/providers/common/digests/sha3_prov.c#L287-L288

paulidale replied:
I don't think they should be hashes, just MACs.  That seems to have slipped through.

slontis replied:
> Specifically, these two lines confused the hell out of me:

Yep, it was done for a reason..

levitte replied:
That makes sense (and helps restore my sanity, thank you :wink:)

levitte replied:
And oh, I did move the KMAC here now...

slontis replied:
Ahh yeah, I must of got confused when I did it...
KMAC/SHA3 use the same KECCAK functionality.

levitte replied:
Ah, I see why it's there...  `crypto/evp/m_sha3.c` implements KMAC as a EVP_MD, because....  well, because what, really?  Was it so it could be used for PRF purposes?

slontis replied:
The single step KDF uses it..

levitte replied:
Aaaaanyway...  I'll have to adjust some things around KMAC.

levitte replied:
Done

mattcaswell commented:
doc-nits is complaining

levitte commented:
> doc-nits is complaining

damnit...

mattcaswell commented on crypto/evp/pkey_mac.c:
```diff
@@ -276,8 +318,18 @@ static int pkey_mac_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
+                OSSL_PARAM_construct_size_t("outlen", &size);
+            params[params_n] = OSSL_PARAM_construct_end();
+
+            return EVP_MAC_CTX_set_params(hctx->ctx, params);
```
Should we check if it was successfully set?

levitte replied:
Hah, that discussion does proliferate into all kinds of places, isn't it?  Yeah, I'll add a check similar to #9562, with a TODO comment to change it when we have a better working mechanism.

levitte replied:
Done

mattcaswell commented on crypto/modes/siv128.c:
```diff
@@ -223,10 +235,11 @@ int CRYPTO_siv128_aad(SIV128_CONTEXT *ctx, const unsigned char *aad,
 
-    mac_ctx = EVP_MAC_CTX_dup(ctx->mac_ctx_init);
-    if (mac_ctx == NULL
+    ;
```
?

levitte replied:
I think I had some debugging code there, and didn't clean it well enough.  Done

mattcaswell commented on providers/default/macs/poly1305_prov.c:
```diff
@@ -0,0 +1,139 @@
+# include <openssl/err.h>
+
+# include "internal/poly1305.h"
+# include "../../../crypto/poly1305/poly1305_local.h"
```
Why is this necessary?

levitte replied:
Because line 42:

``` C
    POLY1305 poly1305;           /* Poly1305 data */
```

richsalz replied:
You really should move the "internal" header file then.  Or turn the POLY135 and SIPHASH declarations into pointers.  Going up and over three levels is wrong: either keep the strutucre opaque or make it available internally.  Won't repeat the comment but it holds for siphash too.

levitte replied:
*Eventually*, I would like to see *all* of poly1305 move to the providers structure.  Alas, this is currently not possible, as there are legacy things in `crypto/poly1305` that we cannot remove, such as `crypto/poly1305/poly1305_ameth.c`, which is solely there to get MAC functionality via EVP_PKEY.

When *all* of poly1305 has moved to providers, I fully expect poly1305_local.h to move along with it, and that path will return to be just "poly1305_local.h".

I don't see the usefulness in moving this header file to something that is more "shared" (such as to `crypto/include/internal`).  I know that @slontis did this differently, but we disagree on how to do this, and that's ok.

mattcaswell replied:
I thoroughly dislike seeing includes like this...and am not entirely convinced by the argument that "it will get moved eventually"...because such things have a habit of not actually happening and then they stick around for ever more. But I won't go so far as to insist on changing this.

slontis replied:
either way it needs a TODO on it.

levitte replied:
That is very true.  Will add that

levitte replied:
TODO added

mattcaswell commented on providers/default/macs/siphash_prov.c:
```diff
@@ -0,0 +1,156 @@
+# include <openssl/err.h>
+
+# include "internal/siphash.h"
+# include "../../../crypto/siphash/siphash_local.h"
```
Again - why is this necessary?

levitte replied:
Because line 43:

``` C
    SIPHASH siphash;             /* Siphash data */
```

slontis replied:
Which is why i moved some headers and TODO'd them when doing the digest work.

levitte replied:
I do remember the dispute we had ;-)

levitte replied:
TODO added

levitte commented on apps/list.c:
```diff
@@ -127,6 +127,7 @@ static void list_digests(void)
     sk_EVP_MD_pop_free(digests, EVP_MD_meth_free);
 }
 
+#if 0                            /* Temporarly disabled */
```
Because #9576, and I'm preparing another branch to implement the MAC algorithm listing based on that.

levitte commented on crypto/evp/mac_lib.c:
```diff
@@ -11,25 +11,19 @@
-}
-
-EVP_MAC_CTX *EVP_MAC_CTX_new(const EVP_MAC *mac)
+EVP_MAC_CTX *EVP_MAC_CTX_new(EVP_MAC *mac)
```
1.  This is *not* a legacy function.

levitte replied:
2.  Maybe we should do that for other fetchable operations...

mattcaswell replied:
I know its not legacy. I was just debating whether or not we should up-ref. On the one hand similar legacy operations don't do it. On the other hand this isn't legacy so maybe we should do it.

levitte replied:
Legacy implementations are constant, they don't go away, i.e. no need for refcounting.
fetched implementations, on the other hand, are not constant, they can be freed, and an inopportune (and erroneous, I agree) free may cause all kinds of havoc with allocated memory possibly not being allocated any more, or allocated for other stuff.  Refcounting removes that potential chaos.

mattcaswell replied:
...and yet there is no up-ref of `mac` in this function...

levitte replied:
Oh... I thought I had...

levitte replied:
Done

levitte commented on apps/mac.c:
```diff
@@ -161,7 +175,7 @@ int mac_main(int argc, char **argv)
     }
 
-    if (!EVP_MAC_final(ctx, NULL, &len)) {
+    if (!EVP_MAC_final(ctx, NULL, &len, BUFSIZE)) {
```
Yeah I guess...  it's not terribly important, `EVP_MAC_final()` ignores it when the buffer is NULL.  But setting it to zero is neater fer sure, so yeah, will fix.

richsalz commented:
> did this differently, but we disagree on how to do this, and that's ok.

No, it's really not okay.  It might be tolerable, but to say that doing the same kind of thing in multiple ways is evidence of bad things.  There is more than one way to do it is Perl's motto, it shouldn't be acceptable to a single C-based project.

Shrug.  My opinion, I don't vote, contents may have shifted in transport, etc., etc.

levitte commented:
Well, this is what happens when a unit gets split into two halves.

So do we want to have POLY1305_CTX and SIPHASH_CTX shared with the rest of libcrypto?  Is that better?  If that is what the rest of y'all prefer, I'll do so, although grudgingly.

slontis commented on crypto/crmf/crmf_pbm.c:
```diff
@@ -193,17 +199,22 @@ int OSSL_CRMF_pbm_new(const OSSL_CRMF_PBMPARAMETER *pbmp,
+                                         strlen(mdname) + 1);
+    macparams[1] =
+        OSSL_PARAM_construct_octet_string("key", basekey, bklen);
+    if ((mac = EVP_MAC_fetch(NULL, "HMAC", NULL)) == NULL
```
macparams[2] is not set??

levitte replied:
Fixed

slontis commented on crypto/err/openssl.txt:
```diff
@@ -1156,6 +1152,11 @@ PROV_F_AES_EINIT:109:aes_einit
 PROV_F_AES_INIT_KEY:110:aes_init_key
 PROV_F_AES_STREAM_UPDATE:111:aes_stream_update
 PROV_F_AES_T4_INIT_KEY:112:aes_t4_init_key
+PROV_F_BLAKE2_MAC_INIT:115:blake2_mac_init
```
Why are we propagating these now?

levitte replied:
`util/mkerr.pl` made me do it.

slontis replied:
You know what they say.. A good workman never blames his tools.. :)

slontis commented on crypto/evp/mac_lib.c:
```diff
@@ -97,101 +107,31 @@ int EVP_MAC_update(EVP_MAC_CTX *ctx, const unsigned char *data, size_t datalen)
-    }
-
-    return ok;
+    return mac->get_params(params);
```
This is assuming the function exists always?

levitte replied:
Good point.

slontis commented on crypto/evp/mac_lib.c:
```diff
@@ -97,101 +107,31 @@ int EVP_MAC_update(EVP_MAC_CTX *ctx, const unsigned char *data, size_t datalen)
-        rv = EVP_MAC_ctrl(ctx, cmd, bin, (size_t)binlen);
-    OPENSSL_free(bin);
-    return rv;
+    return ctx->meth->ctx_get_params(ctx->data, params);
```
ditto

levitte replied:
Fixed too

slontis commented on crypto/evp/mac_lib.c:
```diff
@@ -97,101 +107,31 @@ int EVP_MAC_update(EVP_MAC_CTX *ctx, const unsigned char *data, size_t datalen)
+int EVP_MAC_CTX_set_params(EVP_MAC_CTX *ctx, const OSSL_PARAM params[])
 {
-    return mac->type;
+    return ctx->meth->ctx_set_params(ctx->data, params);
```
ditto

levitte replied:
Yup

slontis commented on crypto/kdf/sskdf.c:
```diff
@@ -40,12 +40,13 @@
 
 struct evp_kdf_impl_st {
-    const EVP_MAC *mac; /* H(x) = HMAC_hash OR H(x) = KMAC */
+    EVP_MAC *mac; /* H(x) = HMAC_hash OR H(x) = KMAC */
```
NIT: realign the comment

levitte replied:
Okie

slontis commented on crypto/kdf/sskdf.c:
```diff
@@ -205,11 +216,18 @@ static int SSKDF_mac_kdm(const EVP_MAC *kdf_mac, const EVP_MD *hmac_md,
+    if (hmac_md != NULL) {
+        const char *mdname = EVP_MD_name(hmac_md);
+        params[params_n++] =
+            OSSL_PARAM_construct_utf8_string("algorithm", (char *)mdname,
```
no hardwired names please

levitte replied:
Darn, I thought I had caught them all

slontis commented on crypto/kdf/sskdf.c:
```diff
@@ -205,11 +216,18 @@ static int SSKDF_mac_kdm(const EVP_MAC *kdf_mac, const EVP_MD *hmac_md,
+                                             strlen(mdname) + 1);
+    }
+    params[params_n++] =
+        OSSL_PARAM_construct_octet_string("key", (void *)salt, salt_len);
```
same here

levitte replied:
Gotcha

slontis commented on crypto/kdf/sskdf.c:
```diff
@@ -430,11 +454,16 @@ static int sskdf_derive(EVP_KDF_IMPL *impl, unsigned char *key, size_t keylen)
+         * block size?
+         */
+        macname = EVP_MAC_name(impl->mac);
+        if (strcmp(macname, "HMAC") == 0) {
```
Magic strings..

levitte replied:
Ok, so what do you propose?

slontis replied:
#define in a header, I guess we have this all over the place..

levitte replied:
I was thinking more in details.  What do I call that define?  Where do I put it, `core_names.h`?

slontis replied:
Sounds good to me.. We can always move it later if needed.

levitte replied:
Do you mind if that happens in a separate PR?  We have *a lot* of places where we use raw name strings...

slontis replied:
No I dont mind..

slontis commented on crypto/kdf/tls1_prf.c:
```diff
@@ -232,19 +233,29 @@ static int tls1_prf_P_hash(const EVP_MD *md,
-    if (EVP_MAC_ctrl(ctx_init, EVP_MAC_CTRL_SET_KEY, sec, sec_len) != 1)
+
+    /* TODO(3.0) rethink "flags", also see hmac.c in providers */
+    mac_flags = EVP_MD_CTX_FLAG_NON_FIPS_ALLOW;
```
This particular flag is a leftover from the old fips module

levitte replied:
Does that mean I can throw this code away?

slontis replied:
I would probably leave it for now - the todo is there.

slontis commented on crypto/modes/siv128.c:
```diff
@@ -167,23 +168,34 @@ int CRYPTO_siv128_init(SIV128_CONTEXT *ctx, const unsigned char *key, int klen,
+    OSSL_PARAM params[3];
+    const char *cbc_name = EVP_CIPHER_name(cbc);
+
+    params[0] = OSSL_PARAM_construct_utf8_string("algorithm",
```
more embedded strings

levitte replied:
Fixed

slontis commented on doc/man3/EVP_MAC.pod:
```diff
@@ -300,14 +272,25 @@ otherwise B<NULL>.
+
+      if (cipher != NULL)
+          params[params_n++] =
+              OSSL_PARAM_construct_utf8_string("cipher", cipher,
```
Not sure we should be using raw strings everywhere.. would prefer defines..

richsalz replied:
if the name is architecturally defined, that having things like this
```
#define CIPHER_PARAM_NAME "cipher"
```
adds a level of indirection and makes things less readable.

paulidale replied:
I don't mind which we use, but we should at least try to be consistent about it.  Thus far, we've been using defines.  Personally, I'd prefer strings but not enough to argue for them.

One distinct advantage of using defines is that typos are detected at compile time, rather than run time.

richsalz replied:
I don't see the advantage:
```
#define CIPHER_PARAM_NAME "cypher"
```

What am I missing?

slontis replied:
Its wrong in one place :)

slontis replied:
I dont care either - it just needs to be consistent. Matt is doing the opposite.

levitte replied:
This is the user manual.  Leave the choice to the user.

slontis commented on providers/common/macs/gmac_prov.c:
```diff
@@ -0,0 +1,289 @@
+    /*
+     * Conditions for legacy EVP_CIPHER uses.
+     */
+    ENGINE *engine;              /* Engine implementating the algorithm */
```
NIT:  implementing

levitte replied:
Fixed

slontis commented on providers/common/macs/gmac_prov.c:
```diff
@@ -0,0 +1,289 @@
+    EVP_CIPHER_CTX *ctx;         /* Cipher context */
+
+    /*
+     * References to the underlying cipher implementation.  tmpcipher
```
NIT: tmpcipher?

levitte replied:
Fixed

slontis commented on providers/common/macs/gmac_prov.c:
```diff
@@ -0,0 +1,289 @@
+     * References to the underlying cipher implementation.  tmpcipher
+     * caches the cipher, always.  alloc_cipher only holds a reference
+     * to an explicitly fetched cipher.
+     * tmpcipher is cleared after a CMAC_Init call.
```
also here

levitte replied:
Yup

slontis commented on providers/common/macs/gmac_prov.c:
```diff
@@ -0,0 +1,289 @@
+    return 1;
+}
+
+static const OSSL_PARAM known_ctx_set_params[] = {
```
I am not sure if these tables are a great idea..

paulidale replied:
These are for the type query function.  Why wouldn't they be a good idea?

slontis replied:
I just dont want people thinking that this is the solution to the setparams issue returning errors or ignoring - it would be ugly for that.

paulidale replied:
Agreed, but I do think they are beneficial overall.

levitte replied:
See #9576 for another kind of usefulness

slontis commented on providers/common/macs/gmac_prov.c:
```diff
@@ -0,0 +1,289 @@
+        if (p->data_type != OSSL_PARAM_OCTET_STRING)
+            return 0;
+
+        if (p->data_size != (size_t)EVP_CIPHER_CTX_key_length(ctx)) {
```
Are we exposing this function to the FIPS module?

levitte replied:
Yup.  It's implemented in `crypto/evp/evp_lib.c`, which is used in the FIPS module.

Also, we build all modules with `-z defs` on Linux (and others too, but Linux is what I remember), so if a function isn't exposed to the FIPS module, the FIPS module build will fail.

slontis commented on test/recipes/30-test_evp_data/evpmac.txt:
```diff
@@ -157,7 +157,7 @@ Output = 5150d1772f50834a503e069a973fbd7c
 Ctrl = digestsize:13
 Key = 000102030405060708090A0B0C0D0E0F
-Result = MAC_CTRL_ERROR
+Result = MAC_BAD_PARAMS
```
Is this a breaking change?

levitte replied:
No.  "MAC_BAD_PARAMS" is a string generated by `test/evp_test.c`

slontis commented on doc/man7/provider-mac.pod:
```diff
@@ -0,0 +1,249 @@
+OP_mac_get_param_types(), OP_mac_ctx_get_param_types(), and
+OP_mac_ctx_set_param_types() all return constant B<OSSL_PARAM> arrays
+as descriptors of the parameters that OSSL_mac_get_params(),
+OSSL_mac_ctx_get_params(), and OSSL_mac_ctx_set_params() can handle,
```
Do you mean OP_mac_ctx_set_params() ? This is the only mention of this name in thhis pod file??

levitte replied:
Oops!  Yes, I do.  Darn slippery fingers :wink:

slontis commented:
I assume you have seen the sha3_prov.c conflict..

levitte commented:
> I assume you have seen the sha3_prov.c conflict..

I have now, and I'm backing out of that change anyway.

levitte commented:
Regarding KMAC as a hash vs KMAC as a MAC, I chose to keep the hash implementation and simply use it from the MAC implementation.  That means the EVP digest functionality can still be used, which makes life easy.

That means we have KMAC128 and KMAC256 as an EVP_MD *and* as an EVP_MAC.  This works technically, since we have separated identity spaces for different operations, but might seem confusing (it certainly was to me for a bit).  On the other hand, NIST calls it a hash algo, at least if you look [here](https://csrc.nist.gov/projects/computer-security-objects-register/algorithm-registration)...

levitte commented:
Remind me, should I move CMAC to providers/common?

levitte commented on doc/man3/EVP_MAC.pod:
```diff
@@ -85,178 +96,162 @@ through diverse controls.
+the memory pointed at by I<out>, and sets its size in the B<size_t>
+the I<outl> points at.
+If I<out> is B<NULL> or I<outsize> is too small, then no computation
+is made.
```
Done

slontis commented:
> That means we have KMAC128 and KMAC256 as an EVP_MD _and_ as an EVP_MAC. This works

We could call it something like KECCAK_KMAC128 (Because that is what it is)..

levitte commented:
> We could call it something like KECCAK_KMAC128 (Because that is what it is)..

Which one?

slontis commented:
> Which one?

The digest one..

levitte commented:
Sure, I can do that.

levitte committed:
Rename the hash implementations KMAC{128,256} to KECCAK_KMAC{128,256}

This avoids getting them confused with the MAC implementations.

slontis commented:
> Sure, I can do that.

Could be a seperate PR if you want to get this PR in..

levitte commented:
> Could be a seperate PR if you want to get this PR in..

It was 6 lines to change in total.  No biggie...

paulidale commented:
> Remind me, should I move CMAC to providers/common?

I'd like to see this happen.  It will eventually be needed in the FIPS provider.

levitte commented:
> I'd like to see this [CMAC] happen. It will eventually be needed in the FIPS provider.

Okie, will do.  It's a bit more involved, as this touches `crypto/cmac`

levitte commented:
Actually, we could opt to make to move of CMAC to the FIPS module a separate PR.  Do you mind?

paulidale commented:
Sure a new PR makes sense.

mattcaswell commented:
Travis failures look relevant

levitte commented:
Will fix this evening

mattcaswell commented on providers/common/macs/kmac_prov.c:
```diff
@@ -123,65 +147,78 @@ static void kmac_free(EVP_MAC_IMPL *kctx)
-    gdst->xof_mode = gsrc->xof_mode;
-    memcpy(gdst->key, gsrc->key, gsrc->key_len);
-    memcpy(gdst->custom, gsrc->custom, gdst->custom_len);
+    dst->md = src->md;
```
Up ref?

levitte replied:
Oh, turns out KMAC wasn't done quite right...

mattcaswell commented on providers/default/macs/blake2_mac_impl.c:
```diff
@@ -0,0 +1,230 @@
+}
+
+/*
+ * ALL parameters should be set before init().
```
Is this a general rule we should be documenting - or something specific to blake2 (which should also be documented somewhere)?

levitte replied:
This is blake2 specific...

mattcaswell commented on providers/default/macs/cmac_prov.c:
```diff
@@ -0,0 +1,261 @@
+
+    dst->tmpengine = src->tmpengine;
+    dst->tmpcipher = src->tmpcipher;
+    dst->alloc_cipher = src->alloc_cipher;
```
Up ref?

levitte replied:
Done

levitte replied:
I'm finding more bugs, so there's more to adjust...

mattcaswell reviewed (changes requested):
Took another pass through this.

levitte commented on doc/man3/EVP_MAC.pod:
```diff
@@ -85,178 +96,156 @@ through diverse controls.
+or SHAKE256.
 
-B<EVP_MAC_CTRL_SET_CIPHER> takes one argument: C<EVP_CIPHER *>
+Also note that "algorithm" can be use generically instead of "md",
```
Have a look at the EVP_PKEY implementations in 1.1.1, and their different control command strings. The variety is huge

levitte commented on crypto/evp/keymgmt_meth.c:
```diff
@@ -44,6 +44,7 @@ static void *keymgmt_from_dispatch(const char *name, const OSSL_DISPATCH *fns,
         return NULL;
     }
 
+
```
Right...  I think this is left-over from work that I broke out to another PR

levitte commented on crypto/evp/mac_lib.c:
```diff
@@ -11,26 +11,24 @@
+        || !EVP_MAC_up_ref(mac)) {
         EVPerr(EVP_F_EVP_MAC_CTX_NEW, ERR_R_MALLOC_FAILURE);
+        if (ctx != NULL)
+            mac->freectx(ctx->data);
```
I've assumed that `freectx` is sane and accepts NULL as input (and acts as a no-op).  Under that condition, this works.

mattcaswell replied:
Yeah, you're right.

levitte commented on crypto/evp/mac_lib.c:
```diff
@@ -41,9 +39,11 @@ EVP_MAC_CTX *EVP_MAC_CTX_new(const EVP_MAC *mac)
-    if (ctx != NULL && ctx->data != NULL) {
-        ctx->meth->free(ctx->data);
+    if (ctx != NULL) {
+        ctx->meth->freectx(ctx->data);
```
Good point.  I'll do it for MACs, but that needs to be said for the other operation types as well...

levitte commented on crypto/evp/mac_lib.c:
```diff
@@ -97,101 +117,43 @@ int EVP_MAC_update(EVP_MAC_CTX *ctx, const unsigned char *data, size_t datalen)
 
-int EVP_MAC_str2ctrl(EVP_MAC_CTX *ctx, int cmd, const char *value)
+/*
+ * The {get,sset}_params functions return 1 if there is no corresponding
```
Fixed

levitte commented on crypto/init.c:
```diff
@@ -558,15 +538,6 @@ int OPENSSL_init_crypto(uint64_t opts, const OPENSSL_INIT_SETTINGS *settings)
             && !RUN_ONCE(&add_all_digests, ossl_init_add_all_digests))
         return 0;
 
-    if ((opts & OPENSSL_INIT_NO_ADD_ALL_MACS)
```
Right, and `OPENSSL_INIT_ADD_ALL_MACS` too

levitte commented on doc/man3/EVP_MAC.pod:
```diff
@@ -60,23 +63,31 @@ B<EVP_MAC_CTX> is a context type that holds internal MAC information
 
+=head2 Algorithm implementation fetching
+
+EVP_MAC_fetch() fetches an implementation of a MAC I<algorithm>, given
```
Okie

levitte commented on doc/man3/EVP_MAC.pod:
```diff
@@ -85,178 +96,156 @@ through diverse controls.
 =over 4
 
-=item B<EVP_MAC_CTRL_SET_KEY>
+=item "key"
```
Agreed

levitte replied:
... actually, I think both should the name and the string should be presented.  I don't see the point with hiding the string value from the users, and let them choose how they want to do things.

levitte commented on doc/man3/EVP_MAC.pod:
```diff
@@ -60,23 +63,31 @@ B<EVP_MAC_CTX> is a context type that holds internal MAC information
+EVP_MAC_up_ref() increments the reference count of an already fetched
+MAC.
+
+EVP_MAC_free() frees a fetched algorithm.
```
Added

levitte commented on doc/man7/provider-mac.pod:
```diff
@@ -0,0 +1,249 @@
+OP_mac_final() completes the MAC computation started through previous
+OP_mac_init() and OP_mac_update() calls.
+The B<mctx> parameter contains a pointer to the provider side context.
+The resulting MACshould be written to B<out> and the amount of data written
```
Fixed

levitte commented on include/openssl/core_numbers.h:
```diff
@@ -217,6 +217,42 @@ OSSL_CORE_MAKE_FUNC(int, OP_cipher_ctx_get_params, (void *cctx,
+# define OSSL_FUNC_MAC_NEWCTX                        1
+# define OSSL_FUNC_MAC_DUPCTX                        2
+# define OSSL_FUNC_MAC_FREECTX                       3
+# define OSSL_FUNC_MAC_INIT                          5
```
It's a magic number, and does a disappear/reappear act :wink:

levitte commented on providers/common/macs/gmac_prov.c:
```diff
@@ -0,0 +1,289 @@
+    }
+
+    dst->cipher = src->cipher;
+    dst->alloc_cipher = src->alloc_cipher;
```
Done

levitte replied:
(I'm actually surprised this didn't bite me ...)

levitte commented on providers/common/macs/hmac_prov.c:
```diff
@@ -0,0 +1,284 @@
+        return NULL;
+    }
+    /* TODO(3.0) Should we do something more with that context? */
+    macctx->provctx = provctx;
```
I'll leave the TODO for now...  we may have reasons to revisit it or remove the TODO later.
To assume the provctx shall be an OPENSSL_CTX is asking for trouble, should it change some time in the future.  *Do not get tempted!*

levitte commented on providers/common/macs/hmac_prov.c:
```diff
@@ -0,0 +1,284 @@
+                EVP_MD_fetch(PROV_LIBRARY_CONTEXT_OF(macctx->provctx),
+                             algoname, propquery);
+
+#if 1       /* TODO(3.0) remove when provider digests are sane */
```
This was from before #9391 went in.  I'll adjust.

levitte commented:
A rebase was needed too, to align this with recent EVP_CIPHER and EVP_MD changes

levitte commented:
CIs are happy again.

levitte commented:
As soon as the discussion on gettable,settable function naming has settled in #9576 and #9591, I will change this branch to match

levitte commented:
#9576 is now merged, I would say that establishes the `[gs]ettable_params` naming standard well enough, so I've done the same here.

levitte commented:
... oh, and rebased, of coursse

levitte commented:
Also, now that #9576 put in place the foundation for listing provider implementations with `openssl list`, I also added a MAC adaptation (which takes away the `#if 0` that I think @mattcaswell was wondering about...)

slontis commented:
doc-nits issues

levitte commented:
> doc-nits issues

Fixed

slontis reviewed (approved):
Good work Richard.
Assuming no test failures I approve.

levitte commented:
> Good work Richard.

Thank you :-)

levitte commented:
I did a larger squashing job and pushed here to ensure that CIs are still happy.  Feel free to have a new look through to see if I messed up... I don't think I did.  If nothing goes wrong and no one complains, I'll use the approvals made so far.

levitte commented:
Merged.

e74bd29053a543ab4908ae8545b46f2e38c98bab Prepare EVP_MAC infrastructure for moving all MACs to providers
55a0a117e7d911752bc7e3e00a67f7a5ad168159 Move BLAKE2 MACs to the providers
2e5db6ad84ecd80954a66e250eae7d96e4565ea1 Move CMAC to providers
d33313be449741aeb68f40831afece4107467bdf Move GMAC to providers
5183ebdcf5aa49e7a6b065a54575aa5748964736 Move HMAC to providers
e23cda000ef2e54e83da61e102ce6d22ed56721a Move KMAC to providers
4657693d9e53ea45cbb903969370ddf9d331dafd Move SipHash to providers
ae0b6b92038b3f6e3a0e2f354cd900f96bce4d8b Move Poly1305 to providers
6a4f9cd113e7fc0734eb4b62e596488e71961040 Remove init of MACs from EVP
d747fb2ec5ee964e8367e7baec8d499d4832def6 Adapt apps/mac.c to use provider based MACs
776796e81895fc66994a90bb62da4c7f50d34368 Adapt diverse code to provider based MACs.
25446a66b69a28c85d178e4454d2caed75d75293 Adapt the MAC tests, and tests for other things that use EVP_MAC
f73eb733eeeb50df0068d01efaa3221cadb07389 Adjust some provider reason codes
810a1d0320fa1706c722d1cfcae398bfb8219837 OSSL_PARAM_construct_from_text(): handle non-hex octet string input
bb31895d87a57e7bb1d3f0939edc53d674094ea4 Rename the hash implementations KMAC{128,256} to KECCAK_KMAC{128,256}
7dd0f299387bac79c304dc7cb8056cd4684fb91f Add EVP_MAC_provider()
d1cafb083d0cb84b2081dd5ca4ba6bed05b8c6ac Implement EVP_MAC_do_all_ex()
467b8e56753e53ed2b9a9f1ad35970a5ceb837d9 Re-implement 'openssl list -mac-algorithms'

h-vetinari commented:
Should be moved to "Done" in the [project board](https://github.com/openssl/openssl/projects/2).

levitte commented:
Done
